### PR TITLE
libc runtime

### PR DESCRIPTION
**Pull request by haampie**
TODO:

- [x] fix some bug in the concretizer that prevents this
- [x] `libc` automatic external
- [x] make `libc` a virtual, make `gcc-runtime` depend on `libc`
- [x] drop `os` match rules on `linux`
- [x] allow reuse of older libc assuming abi compat
- [x] Ensure every compiler we support on linux can detect libc
- [x] Add unit tests stressing libc compatibility
- [x] Ensure externals on `linux` still work correctly
- [x] postponed: deal with out-of-prefix `PT_INTERP` loader path relocation

### CODE REVIEW

**Comment by alalazo** [comment]
With respect to:

> fix some bug in the concretizer that prevents this (circular dep with gcc-runtime?)

This is working, provided we add / inject `glibc` as an external package. Otherwise there will be circular dependencies due to `glibc` dependency on e.g. `linux-kernel` that depends on `gcc-runtime` and `glibc` itself. 

For instance, on:

* **Spack:** 0.22.0.dev0 (9a17670e61fc283335208385e11096b95d5e7e67)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

and using:
```yaml
packages:
  glibc:                                                                                                                                                                                                                                                                         
    externals:                                                                                                                                                                                                                                                                   
    - spec: "glibc@=2.31"                                                                                                                                                                                                                                                        
      prefix: /usr          
```

I get the following for:
```console
$ spack solve zlib-ng
```

<details>

<summary>zlib DAG</summary>

![zlib](https://github.com/spack/spack/assets/4199709/f87ffcd0-d1e9-46ed-add0-acc38eaa89ad)

</details>

**Comment by haampie** [comment]
Right, my original idea was to make it an automatic external

**Comment by alalazo** [comment]
The last commit (https://github.com/spack/spack/pull/43190/commits/19d2c2c6f70d83d0420ed30ef813ac4f7df9214c) gives a fully working GCC, but we still need some tightening for other compilers. Anyhow, to try the new feature:
```console
$ spack mirror add devtools https://raw.githubusercontent.com/alalazo/test-buildcaches/main/devtools
$ spack buildcache keys -it 
$ spack env activate --temp
$ spack add hdf5~mpi
$ spack concretize --reuse -f
```
On my platform:

* **Spack:** 0.22.0.dev0 (19d2c2c6f70d83d0420ed30ef813ac4f7df9214c)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

This reuses a bunch of nodes from the buildcache that has `almalinux8` builds:

Screenshot from 2024-04-16 18-52-20

**Comment by alalazo** [comment]
The shell test is currently failing, because at the moment they contain some mock installation that do not inject libc - while I started requiring that on linux in https://github.com/spack/spack/pull/43190/commits/ebb113ae3c6b760e62ed87cfe5b14a7f10676d3c

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -2478,6 +2478,23 @@ def define_runtime_constraints(self):
             if hasattr(compiler_cls, "runtime_constraints"):
                 compiler_cls.runtime_constraints(spec=compiler.spec, pkg=recorder)
 
+            # Inject libc from available compilers, on Linux
+            if not compiler.available:
+                continue
+
+            if spack.platforms.host().name == "linux":
```
Comment (haampie) [inline lib/spack/spack/solver/asp.py]: We may also include BSD (later?). Plus at some point this should be a target not host condition :thinking: 

Can this branch already be written in terms of the platform `compiler` targets?

File: lib/spack/spack/solver/asp.py [hunk 2]
```diff
@@ -2427,6 +2478,23 @@ def define_runtime_constraints(self):
...
+                libc = compiler.compiler_obj.default_libc()
+                if libc:
+                    recorder("*").depends_on(
+                        "libc", when=f"%{str(compiler.spec)}", type="link", description="Add libc"
```
Comment (haampie) [inline lib/spack/spack/solver/asp.py]:
```suggestion
                        "libc", when=f"%{compiler.spec}", type="link", description="Add libc"
```

File: lib/spack/spack/solver/asp.py [hunk 3]
```diff
@@ -2427,6 +2478,23 @@ def define_runtime_constraints(self):
...
+                    )
+                    recorder("*").depends_on(
+                        str(libc),
+                        when=f"%{str(compiler.spec)}",
```
Comment (haampie) [inline lib/spack/spack/solver/asp.py]:
```suggestion
                        when=f"%{compiler.spec}",
```

**Comment by haampie** [comment]
```
$ spack spec -L zlib-ng
Input spec
--------------------------------
 -   zlib-ng

Concretized
--------------------------------
 -   gaxnrttfo2lk7yzdd2nsm5yxf55zadsv  zlib-ng@2.1.6%gcc@13.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu23.10-zen2
[+]  tofiopdeqtfqrjqrfnclgnvx2d47yraq      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu23.10-zen2
[e]  gx2zlxihjlhzfkcz47rjr2iir5hrogi7          ^glibc@2.38%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen2
[e]  gx2zlxihjlhzfkcz47rjr2iir5hrogi7      ^glibc@2.38%gcc@13.2.0 build_system=autotools arch=linux-ubuntu23.10-zen2
[+]  boxbsmc7a5ycptkstw777ao5xecdo34h      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-ubuntu23.10-zen2
```

not sure what that is about.

One `glibc` reused, one "fresh" resulting in two instances of the same object?

(Have a hard time reducing it to mwe...)

**Comment by alalazo** [comment]
> not sure what that is about.

The issue was that the libc was not "reused". Fixed in https://github.com/spack/spack/pull/43190/commits/b049375976d012f110becb78824a5df63edaaedc

File: lib/spack/spack/solver/asp.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/solver/asp.py]: Would you be fine with https://github.com/spack/spack/pull/43190/commits/85d8bb10705c613c5e3ca1ffbcde3bbc74c18488 for now (which just adds a function where to encode when libc compatibility is used)?

**Comment by haampie** [comment]
To sum up the relocation struggles:

1. The interpreter path is on most distros "out of prefix", meaning default prefix -> prefix relocation is insufficient
2. On almost all distros it's out of prefix exactly such that relocation is unnecessary
3. *But* we do care about that small fraction of distros that is different (EESSI users w/ Gentoo Prefix, and NixOS)
4. We preferably don't wanna model in terms of *the* system libc, there can be multiple
5. It's most robust to determine the interpreter path from the compiler's link line
6. But we also want `spack install gcc` to work without a compiler installed: there may not be a compiler to query where libc is.
7. If no configured compiler: we can determine the interpreter path from the current Python process's `PT_INTERP` (except if Python is static, which is a valid use case when bootstrapping Spack on systems without Python)
8. Alternatively we could run `/usr/bin/ldd` or `ldd` from PATH, or scan `/usr` for a `libc.so` with arch compatible with the currently running Python process.
9. Finally there's an annoying issue that after concretization we just have `libc` with  say `prefix=/usr`, but not the interpreter path, so we'd have to query the compiler again to get it.

One way or another we get the interpreter path, and add a replacement `<old interpreter path> -> <new interpreter path>`.

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
One test in pipelines took more than 6 hrs. I guess we need to put a timeout on tests.

**Comment by haampie** [comment]
We decided to stick with exact prefix match for now so that reuse between nix / gentoo prefix and classical distros isn't possible. That makes the scope of this pr a bit smaller.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**